From 86cb34ec8e9a736a6134ca41a23fd3151c03c806 Mon Sep 17 00:00:00 2001
From: Till Kamppeter <till.kamppeter@gmail.com>
Date: Tue, 9 Aug 2016 18:11:28 +0200
Subject: Some printers have broken device IDs with newline

 characters inside. These break the cups-deviced printer discovery
 mechanism and so the printers get ignored. This patch allows newline
 characters in device IDs
Bug-Ubuntu: https://bugs.launchpad.net/bugs/468701
Bug: https://www.cups.org/str.php?L4345
Last-Update: 2015-02-10

Patch-Name: cups-deviced-allow-device-ids-with-newline.patch
---
 scheduler/cups-deviced.c | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/scheduler/cups-deviced.c b/scheduler/cups-deviced.c
index 9a1b8d3e1..3cc4eecd0 100644
--- a/scheduler/cups-deviced.c
+++ b/scheduler/cups-deviced.c
@@ -562,15 +562,30 @@ get_device(cupsd_backend_t *backend)	/* I - Backend to read from */
 
     if (*ptr == '\"')
     {
-      for (ptr ++, device_id = ptr; *ptr && *ptr != '\"'; ptr ++)
+      for (ptr ++, device_id = ptr; *ptr != '\"'; ptr ++)
       {
 	if (*ptr == '\\' && ptr[1])
 	  _cups_strcpy(ptr, ptr + 1);
+	if (!*ptr)
+	{
+	  fprintf(stderr, "WARNING: [cups-deviced] Possible newline in device ID \"%s\": %s\n",
+		  backend->name, line);
+	  *ptr = ' ';
+	  ptr ++;
+	  *ptr = 0;
+	  if (!cupsFileGets(backend->pipe, ptr, sizeof(line) - (ptr - temp)))
+	  {
+	    cupsFileClose(backend->pipe);
+	    backend->pipe = NULL;
+	    fprintf(stderr, "ERROR: [cups-deviced] Bad line from \"%s\": %s\n",
+		    backend->name, line);
+	    return (-1);
+	  }
+	}
+	if (!*ptr)
+	  goto error;
       }
 
-      if (*ptr != '\"')
-	goto error;
-
       for (*ptr++ = '\0'; isspace(*ptr & 255); *ptr++ = '\0');
 
      /*
